Mathlib/NumberTheory/Padics/Harmonic.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2023 Koundinya Vajjha. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Koundinya Vajjha
+-/
+
+import Mathlib.NumberTheory.Padics.PadicIntegers
+import Mathlib.Data.Int.Log
+/-!
+
+The nth Harmonic number is not an integer. We formalize the proof using
+2-adic valuations. This proof is due to Kürschák.
+
+Reference:
+https://kconrad.math.uconn.edu/blurbs/gradnumthy/padicharmonicsum.pdf
+
+-/
+
+open BigOperators
+
+/-- The nth-harmonic number defined as a finset sum of consecutive reciprocals. -/
+def harmonic : ℕ → ℚ := fun n => ∑ i in Finset.range n, 1 / (i + 1)
+
+lemma harmonic_pos {n : ℕ} (Hn : n ≠ 0) : 0 < harmonic n :=
+  Finset.sum_pos (fun _ _ => div_pos zero_lt_one (by norm_cast; linarith))
+  (by rwa [Finset.nonempty_range_iff])
+
+lemma harmonic_singleton {n c : ℕ} (hc : c ∈ Finset.range n) :
+    harmonic n = 1 / (c + 1) + ∑ x in Finset.range n \ {c}, 1 / ((x : ℚ) + 1) := by
+  dsimp [harmonic]
+  rwa [add_comm, Finset.sum_eq_sum_diff_singleton_add (i := c)]
+
+lemma finset_range_sdiff_singleton_nonempty {c n : ℕ} (hn : 2 ≤ n) :
+    Finset.Nonempty (Finset.range n \ {c}) := by
```

Comment:
This feels like it should be a quick consequence of two simpler lemmas, one saying `(Finset.range n).Nontrivial` and one saying that `s.Nontrivial → (s \ {c}).Nonempty`. And can you move it to an earlier file?

Reply:
I tried this for a while but gave up. I tried to put this in `Finset.Basic` but the lemmas there don't know `linarith` apparently. This is what I could come up with: 
```lean
lemma finset_range_sdiff_singleton_nonempty {c n : ℕ} (hn : 2 ≤ n) :
    Finset.Nonempty (Finset.range n \ {c}) := by
  rw [Finset.sdiff_nonempty, Finset.subset_singleton_iff, Finset.range_eq_empty_iff, not_or]
  refine' ⟨ne_of_gt <| lt_of_lt_of_le zero_lt_two hn, fun Hnot => _⟩
  have : n = 1 := by rw [← Finset.card_range n, ← Finset.card_singleton c, Hnot]
  simp only [this] at hn
```
Feel free to golf this and commit if you can

Reply:
I mean that you should write two lemmas, one saying `(Finset.range n).Nontrivial` and one saying that `s.Nontrivial → (s \ {c}).Nonempty`. You might need to merge master to get access to `Finset.Nontrivial`.

Reply:
Ah okay, that explains it

Reply:
Ok this is done in 6338778d53fbc18e738d3aa359bcfadad45a506d.